packages/@lwc/engine-core/src/framework/weak-multimap.ts

```diff
@@ -0,0 +1,110 @@
+    private _map: WeakMap<K, Set<V>> = new WeakMap();
+
+    private _getValues(key: K): Set<V> {
+        let values = this._map.get(key);
+        if (isUndefined(values)) {
+            values = new Set();
+            this._map.set(key, values);
+        }
+        return values;
+    }
+
+    get(key: K): ReadonlySet<V> {
+        return this._getValues(key);
+    }
+    add(key: K, vm: V) {
+        const set = this._getValues(key);
+        set.add(vm);
+    }
+    clear(key: K) {
+        this._map.delete(key);
+    }
+}
+
+// This implementation relies on the WeakRef/FinalizationRegistry proposal.
+// For some background, see: https://github.com/tc39/proposal-weakrefs
+class ModernWeakMultiMap<K extends object, V extends object> implements WeakMultiMap<K, V> {
+    private _map = new WeakMap<K, WeakRef<V>[]>();
+
+    private _registry = new FinalizationRegistry((weakRefs: WeakRef<V>[]) => {
+        // This should be considered an optional cleanup method to remove GC'ed values from their respective arrays.
+        // JS VMs are not obligated to call FinalizationRegistry callbacks.
+
+        // Work backwards, removing stale VMs
+        for (let i = weakRefs.length - 1; i >= 0; i--) {
+            const vm = weakRefs[i].deref();
+            if (isUndefined(vm)) {
+                weakRefs.splice(i, 1); // remove
+            }
+        }
+    });
+
+    private _getWeakRefs(key: K): WeakRef<V>[] {
+        let weakRefs = this._map.get(key);
+        if (isUndefined(weakRefs)) {
+            weakRefs = [];
+            this._map.set(key, weakRefs);
+        }
+        return weakRefs;
+    }
+
+    get(key: K): ReadonlySet<V> {
+        const weakRefs = this._getWeakRefs(key);
+        const result = new Set<V>();
+        for (const weakRef of weakRefs) {
+            const vm = weakRef.deref();
+            if (!isUndefined(vm)) {
+                result.add(vm);
+            }
+        }
+        return result;
+    }
+    add(key: K, value: V) {
+        const weakRefs = this._getWeakRefs(key);
+        // We could check for duplicate values here, but it doesn't seem worth it.
+        // We transform the output into a Set anyway
+        weakRefs.push(new WeakRef<V>(value));
+
+        // It's important here not to leak the second argument, which is the "held value." The FinalizationRegistry
+        // effectively creates a strong reference between the first argument (the "target") and the held value. When
+        // the target is GC'ed, the callback is called, and then the held value is GC'ed.
+        // Putting the key here would mean the key is not GC'ed until the value is GC'ed, which defeats the purpose
+        // of the WeakMap. Whereas putting the weakRefs array here is fine, because it doesn't have a strong reference
+        // to anything. See also this example:
+        // https://gist.github.com/nolanlawson/79a3d36e8e6cc25c5048bb17c1795aea
+        this._registry.register(value, weakRefs);
```

Comment:
Realized I had a bug in the weakmap, although it doesn't really matter because the keys were never GC'ed anyway... but it's fixed now.